README.md
[![License](https://img.shields.io/badge/License-BSD_3--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause) ![PyPI](https://img.shields.io/pypi/v/nbi)
[![Tests](https://github.com/kmzzhang/nbi/actions/workflows/tests.yml/badge.svg)](https://github.com/kmzzhang/nbi/actions/workflows/tests.yml)
[![Documentation Status](https://readthedocs.org/projects/nbi/badge/?version=latest)](https://nbi.readthedocs.io/en/latest/?badge=latest)

## nbi: neural bayesian inference

[Documentation](https://nbi.readthedocs.io/en/latest/)

Do you have challenging inference problems that are difficult to solve with standard optimization and/or MCMC methods?
Are you looking to fit the same forward model to thousands or millions of observed targets?
`nbi` may be your solution.

`nbi` is an engine for Neural Posterior Estimation (NPE) focused on out-of-the-box functionality for astronomical data,
particularly light curves and spectra.
`nbi` provides effective embedding/featurizer networks for spectra and light-curve data, along
with importance-sampling integration that enables asymptotically exact inference so that the inference results are
interpretable and trustworthy.

## Installation

You may either install `nbi` with `pip install nbi` or directly from source. As `nbi` is currently under active development,
installing from source may be preferable at this stage.

```bash
git clone https://github.com/kmzzhang/nbi.git
cd nbi
pip install .
```

If you are using Mac ARM **CPU** (i.e. M1/M2/M3), you might want to install PyTorch from source and disable NNPACK, which is known to
reduce performance (see [issue](https://github.com/pytorch/pytorch/issues/107534)). Note that currently the `MPS`
Also support for `weight_norm` on Mac M1-M3 GPUs is recently
[implemented](https://github.com/pytorch/pytorch/pull/108008) but has not been included in a stable
release yet. Installing the nightly version from source also enables `weight_norm` for
the MPS device.

```bash
git clone --recursive https://github.com/pytorch/pytorch
cd pytorch
USE_NNPACK=0 python setup.py install
```

## Quick Start

The `examples/` directory contains complete examples that demonstrates the functionality of `nbi`. A bare-bone
example below illustrates the basic API, which follows the scikit-learn style. The default featurizer network for
sequential data is `resnet-gru`, which is a hybrid CNN-RNN architecture.

Here are a rule of thumb for resnet-gru hyperparameters:

- dim_in: this is your number of input data channels
- depth: number of ResNet blocks. Start near log2(L)-5, where L is length of your sequential data.
- max_hidden: Maximum hidden dimensions for ResNet. Hidden dimensions double (from hidden_conv=32 by default) every
  depth. At least a few times D^2, where D is the dimension of the physical parameter space.

```python
import nbi

# hyperparameters
featurizer = {
    "type": "resnet-gru",
    "dim_in": 1,
    "max_hidden": 64
}

flow = {
    "n_dims": 1,        # parameter space dimension
    "flow_hidden": 32,  # generally no larger than max_hidden
    "num_blocks": 10    # depends on complexity of posterior shape
}

engine = nbi.NBI(
    flow,
    featurizer,
    simulator,
    noise,
    priors,
    device='cpu'        # 'cuda', 'cuda:0', 'mps' for M1/M2 Mac GPU
)
engine.fit(
    n_sims=1000,
    n_rounds=1,
    n_epochs=100
)
y_pred, weights = engine.predict(x_obs, x_err, n_samples=2000)
```

## References

nbi: the Astronomer's Package for Neural Posterior Estimation
([Zhang et al. 2023](https://arxiv.org/abs/2312.03824)). Accepted to NeurIPS 2023 Workshop on Deep Learning and
Inverse Problems.

Masked Autoregressive Flow for Density Estimation (Papamakarios et al. 2017)\
https://arxiv.org/abs/1705.07057

Featurizers: ResNet (He et al. 2015; https://arxiv.org/abs/1512.03385), Gated Recurrent Units
(GRU; Cho et al. 2014; https://arxiv.org/abs/1406.1078),
ResNet-GRU (Zhang et al. 2021; https://iopscience.iop.org/article/10.3847/1538-3881/abf42e)

## Acknowledgments

The `nbi` package is expanded from code originally written for _''Real-time Likelihood-free Inference of Roman Binary Microlensing Events
with Amortized Neural Posterior Estimation'''_ ([Zhang et al. 2021](https://iopscience.iop.org/article/10.3847/1538-3881/abf42e)).
The Masked Autoregressive Flow in this package is partly adapted from the implementation in
https://github.com/kamenbliznashki/normalizing_flows.
Work on this project was supported by the [National Science Foundation award #2206744](https://www.nsf.gov/awardsearch/showAward?AWD_ID=2206744&HistoricalAwards=false) ("CDS&E: Accelerating Astrophysical Insight at Scale with Likelihood-Free Inference").

<center><img src="https://new.nsf.gov/themes/custom/nsf_theme/components/images/logo/logo-desktop.svg" width="30%"></center>


HISTORY.rst
.. :changelog:

0.2.1 (2023-04-20)
++++++++++++++++++

- Clean up redundant arguments
- revised API that follows scikit-learn (engine.fit)
- noise argument support ndarray for iid Gaussian


src/nbi/model.py
from torch import nn

from .nn import RNN, ResNet, flows


class DataParallelFlow(nn.DataParallel):
    def __init__(self, *args, **kwargs):
        super(type(self), self).__init__(*args, **kwargs)

    def sample(self, x, n=1000, is_feature=False):
        if not is_feature:
            cond_vector = self.module.featurizer(x)
        else:
            cond_vector = x
        return self.module.flow.sample(num_samples=n, cond_inputs=cond_vector)


class Flow(nn.Module):
    def __init__(self, featurizer, model):
        super(type(self), self).__init__()
        self.featurizer = featurizer
        self.flow = model

    def forward(
        self,
        x,
        y=None,
        is_feature=False,
        reduce=True,
        return_entropy=False,
        n_entropy=10,
        n=1000,
        aux=None,
        sample=False,
    ):
        if sample:
            if not is_feature:
                cond_vector = self.featurizer(x, aux)
            else:
                cond_vector = x
            return self.flow.sample(num_samples=n, cond_inputs=cond_vector)

        if not is_feature:
            cond_vector = self.featurizer(x, aux)
        else:
            cond_vector = x
        self.cond_vector = cond_vector
        neg_log_probs = -1 * self.flow.log_probs(y, cond_vector)
        if reduce:
            neg_log_probs = neg_log_probs.sum(-1, keepdim=True)
        if not return_entropy:
            return neg_log_probs
        else:
            entropy = self.flow.entropy(num_samples=n_entropy, cond_inputs=cond_vector)
            return neg_log_probs, entropy


def get_featurizer(network_type, config):
    if network_type == "resnet-gru":
        return ResNet(
            config["dim_in"],
            config.pop("dim_out", -1),
            depth=config["depth"],
            kernel_size=config.pop("kernel", 3),
            hidden_conv=config.pop("dim_conv_min", 32),
            max_hidden=config.pop("dim_conv_max", 256),
            norm=config.pop("norm", "weight_norm"),
            rnn_layer=config.pop("n_rnn", 2),
        )
    elif network_type == "resnet":
        return ResNet(
            config["dim_in"],
            config.pop("dim_out", -1),
            depth=config["depth"],
            kernel_size=config.pop("kernel", 3),
            hidden_conv=config.pop("dim_conv_min", 32),
            max_hidden=config.pop("dim_conv_max", 256),
            norm=config.pop("norm", "weight_norm"),
            rnn_layer=0,
        )
    elif network_type == "gru":
        return RNN(
            config["dim_in"],
            hidden_rnn=config["dim_hidden"],
            num_layers=config["depth"],
            num_class=config["dim_out"],
            hidden=config["dim_out"],
            dropout_rnn=0.15,
            bidirectional=False,
            rnn="GRU",
        )


def get_flow(
    featurizer,
    n_dims,
    flow_hidden,
    num_cond_inputs,
    num_blocks=5,
    perm_seed=0,
    clamp_0=-1,
    clamp_1=1,
    n_mog=8,
):
    modules = []
    MADE = flows.MADE2
    num_blocks -= 1

    for i, _ in enumerate(range(num_blocks)):
        modules += [
            flows.Shuffle(n_dims, perm_seed + i),
            MADE(
                n_dims,
                flow_hidden,
                num_cond_inputs,
                shift_only=False,
                linear_scale=False,
                clamp_0=clamp_0,
                clamp_1=clamp_1,
            ),
        ]

    modules += [
        flows.Shuffle(n_dims, perm_seed + num_blocks + 1),
        flows.MADEMOG(
            n_dims,
            flow_hidden,
            num_cond_inputs,
            n_components=n_mog,
            shift_only=False,
            linear_scale=False,
            clamp_0=clamp_0,
            clamp_1=clamp_1,
        ),
    ]
    flow = flows.FlowSequentialMOG(*modules)
    flow.init(n_mog)
    flow.set_num_inputs(n_dims)

    for module in flow.modules():
        if isinstance(module, nn.Linear):
            nn.init.orthogonal_(module.weight)
            if hasattr(module, "bias") and module.bias is not None:
                module.bias.data.fill_(0)

    full_model = Flow(featurizer, flow)
    return full_model
